File: algorithms/tree_node.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def make_tree(nodes: List[int]) -> List[TreeNode]:
    v = []
    for n in nodes:
        v.append(TreeNode(n) if n is not None else None)

    size = len(v) - 1
    for i in range(len(v)):
        n = 2 * i + 1
        if n > size:
            break

        if v[i] is not None:
            v[i].left = v[n]
            n += 1
            if n > size:
                break
            v[i].right = v[n]

    return v

File: algorithms/tree_node.cpp
#include "tree_node.hpp"

namespace tree
{
std::vector<std::unique_ptr<TreeNode>> make_tree(std::initializer_list<std::optional<int>> node_values)
{
    std::vector<std::unique_ptr<TreeNode>> v;

    for (auto opt : node_values)
    {
        v.emplace_back(opt ? std::make_unique<TreeNode>(opt.value()) : std::unique_ptr<TreeNode>());
    }

    const auto size = v.size() - 1;

    for (std::size_t i = 0; i < size; i++)
    {
        std::size_t n = 2 * i + 1;
        if (n > size)
            break;

        if (v[i])
        {
            v[i]->left = v[n++].get();
            if (n > size)
            {
                break;
            }
            v[i]->right = v[n].get();
        }
    }

    return v;
}
} // namespace tree
